modules/standard/Socket.md

```diff
@@ -0,0 +1,210 @@
+# Socket Module API
+
+To those familiar with the Unix socket API, the method names will feel familiar, though their usage will be somewhat simpler than the raw Unix socket API. The Module will comprise of four records which will abstract lower order functions to higher and more intutive methods for user:
+- ipAddr : Address information storage structure.
+- tcpConn : A TCP Socket with established connection
+- tcpListener : A TCP Socket listening for connections
+- udpSocket : A connection less UDP Socket
+
+## Address Information
+
+`ipAddr` serves the purpose of storing address information in a single record irrespective of the protocol used which varies the buffer required to store address.
+
+To initialize an `ipAddr` there are several available overloads and defaults.
+
+**Method**
+```python
+proc ipAddr.init(host:string = "localhost", port:uint(16) = 8000, family:IPFamily = IPFamily.IPv4)
```

Comment:
Should `family` be param?

Reply:
without family, it won't be possible to distinguish btw IPv4 and IPv6 addresses which have different structures something like `127.0.0.1` to `::1` which are both `localhost` but one is ipv4 and the other is ipv6.

Reply:
Right, but what I'm asking is whether it should be a compile-time value:

```chapel
proc ipAddr.init(host:string = "localhost", port:uint(16) = 8000, param family:IPFamily = IPFamily.IPv4)
```

This will make an ipv4 address be a different type than an ipv6 address, and cross-type operations will be subject to the limitations/checks that our type system imposes. That seems like an added benefit to me, but maybe we want to use v4 and v6 addresses more interchangeably. And that part is beyond my expertise on the subject matter.

Reply:
In networking programs, it's common to e.g. accept an IP address to work with from the command line and then to work with both ipv4 and ipv6. Additionally, DNS resolution can give a combination of ipv4 and ipv6 results. Lastly, the OS calls we are interacting with are already runtime polymorphic in this way. So I think it's to our advantage *not* to make ipv4 vs ipv6 a compile-time property of an address.

Reply:
Thanks for the background Michael. All that makes sense to me.